### PR TITLE
Validate Input

### LINKED ISSUE #6: 6. As a user, I want to be alerted when I’m entering an item that is the same as something already on my list so I can avoid duplicate items.
AC:

* Show an error message if the user tries to submit a new item that has the exact same name as an existing item
* Show an error message if the user tries to submit a new item that has the same name as an existing item, where capitalization has been normalized and punctuation has been removed
* The user’s original input should be what gets saved in the database

### PR DESCRIPTION
**Pull request by lena-A-Al**
## Description

Added logic to check if the item that the user inputs is already on the list, and if so, do not add it to the list to avoid duplicates. This validation check normalizes capitalization and removes punctuation/special characters.

We added a new state `message` to set the message to display to the user if the item was added successfully, or if it was not due to the item already being added to the list.

## Related Issue

closes #6

## Acceptance Criteria

- [x] Show an error message if the user tries to submit a new item that has the exact same name as an existing item
- [x] Show an error message if the user tries to submit a new item that has the same name as an existing item, where capitalization has been normalized and punctuation has been removed
- [x] The user’s original input should be what gets saved in the database

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
|     | :hammer: Refactoring       |
|     | :100: Add tests            |
|    | :link: Update dependencies |
|     | :scroll: Docs              |

### Before
Screen Shot 2022-04-16 at 11 53 20 AM

### After
Screen Shot 2022-04-21 at 6 49 23 PM
Screen Shot 2022-04-21 at 6 49 33 PM

## Testing Steps / QA Criteria

- From your terminal, pull down this branch with git pull origin rt-la-validate-input and check that branch out with git checkout  rt-la-validate-input 
- Run npm start to launch the app
- Click the link to "Add Item" at the bottom of the page. Complete the form and click on "Add Item." You should see a success message, and if you switch to "Shopping List" in the navigation links, you should see your item added
- Then try to add that same item but with capitalization changes, or by adding punctuation/special characters. You should receive a message indicating that this item has already been added. Navigate to "Shopping List" in the navigation links, and you should NOT see your item that was inputted with the capitalization and/or punctuation/special characters added to the list.

### CODE REVIEW

File: src/components/AddItem.jsx [hunk 1]
```diff
@@ -1,27 +1,51 @@
...
+    });
+    return () => {
+      unsubscribe();
+    };
+  }, [userToken]);
 
   const submitHandler = async (e) => {
     e.preventDefault();
 
+    const regex = /[!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~]/g;
```
Comment (adidalal) [inline src/components/AddItem.jsx]: Reminds me of [this XKCD](https://xkcd.com/208/)

regex

Comment (adidalal) [reply src/components/AddItem.jsx]: I would suggest inverting this logic

```
var input = 'Item -_*(!@with     #$%^&*()random_-={}[]:\"chars<>,.?/~`';
var normalized_string = input.replace(/[^a-zA-Z\s]/g, '').replace(/\s+/g, ' ').toLowerCase();
console.log(normalized_string);
```

breaking it down:
 - we want to replace all non alphabet characters and spaces with nothing
 - then replace multiple spaces with a single space
 - then lowercase the whole thing
 
 https://regex101.com is useful for this kind of thing.

Comment (rachael-t) [reply src/components/AddItem.jsx]: Thank you for that suggestion! That makes sense, and I've updated the code for that.

**Review by simonkaine — COMMENTED** [review]
I've never used Regex before but I can accurately follow the flow of what you're trying to accomplish here. Seems to be very quick and efficient syntax for checking for misuses here. I've followed your testing steps and everything seems to be working just as expected. Awesome job! 

Minor edit suggestion: I believe you've mislabeled the branch name in the testing steps. Took me a second to figure out that it was meant to say 'input' instead of 'form' when trying to pull the branch.

File: src/components/AddItem.jsx [hunk 2]
```diff
@@ -1,27 +1,58 @@
...
+    const cleanList = docs.map((listItem) =>
+      listItem
+        .data()
+        .item.replace(/[^a-zA-Z\s]/g, '')
+        .replace(/\s+/g, ' ')
+        .toLowerCase(),
+    );
+
+    const cleanItemName = itemName
+      .replace(/[^a-zA-Z\s]/g, '')
```
Comment (adidalal) [inline src/components/AddItem.jsx]: optional suggestion/minor: refactor this to a helper method, as I notice the same "clean" chain applied in two spots

**Review by adidalal — APPROVED** [review]
Nice work!

**Comment by nikkigraybeal** [comment]
Nice work! I did some testing with different inputs and something isn't working correctly:
Screen Shot 2022-04-23 at 8 09 06 AM

Your code looks super close!  It looks to me like you are replacing inner spaces with a single space on lines 32 and 38? Changing that to an empty string seems to fix it, but I'm not super clear how/why! 😆 

Screen Shot 2022-04-23 at 8 15 16 AM

**Comment by rachael-t** [comment]
> Minor edit suggestion: I believe you've mislabeled the branch name in the testing steps. Took me a second to figure out that it was meant to say 'input' instead of 'form' when trying to pull the branch.

Whoops! Thanks for catching that @simonkaine - updated! 😅

File: src/components/AddItem.jsx [hunk 3]
```diff
@@ -1,27 +1,58 @@
...
+  }, [userToken]);
 
   const submitHandler = async (e) => {
     e.preventDefault();
 
+    const cleanList = docs.map((listItem) =>
+      listItem
+        .data()
+        .item.replace(/[^a-zA-Z\s]/g, '')
+        .replace(/\s+/g, '')
```
Comment (rachael-t) [inline src/components/AddItem.jsx:31]: @nikkigraybeal I just updated lines 32 and 38 to replace them with an empty string and I think that fixes the issue? Can you test on your end as well?

Comment (nikkigraybeal) [reply src/components/AddItem.jsx:31]: yes! I tested with that fix and it seems to work no matter what spaces and special chars I include 🎉

**Review by nikkigraybeal — APPROVED** [review]

File: src/components/AddItem.jsx (hunk 2 above)
Comment (rachael-t) [reply src/components/AddItem.jsx]: Thanks! That's a great call-out. I've refactored this to help "DRY" up the code a bit.

**Comment by adidalal** [comment]
@nikkigraybeal Here's the explanation for why replacing multiple spaces with a single one lets you add to the list:
```
H!a. M -> Ha M -> ha m
H! a m. . -> H a m -> h a m
ham -> ham -> ham
```
Those three are different.

Compare this to:
```
H!a. M -> Ha M -> ham
H! a m. . -> H a m -> ham
ham -> ham -> ham
```
I missed the part where we check equality without all punctuation (and spaces are indeed punctuation!) **So, you're absolutely correct.**

Fun note, you can actually simplify this from `.replace(/[^a-zA-Z\s]/g, '').replace(/\s+/g, '')` to just `.replace(/[^a-zA-Z]/g, '')`. (no need for the chained replace, just strip all whitespace in the initial replace). 

I'll go ahead and merge this in.

P.S. A final bit of trivia - `\w` matches any letter, digit or underscore, which is equivalent to `[a-zA-Z0-9_]`

**Pull request merged**